File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/quilt/QuiltDownloadInfo.java
package org.wdt.wdtc.core.download.quilt;

import com.google.gson.JsonObject;
import lombok.Getter;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.ModUtils;

import java.io.File;
import java.io.IOException;

public class QuiltDownloadInfo implements ModDownloadInfoInterface {
  private static final String LibraryListUrl = "https://meta.quiltmc.org/v3/versions/loader/%s/%s/profile/json";

  protected final Launcher launcher;
  @Getter
  protected final String QuiltVersionNumber;

  public QuiltDownloadInfo(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
  }

  public QuiltDownloadInfo(Launcher launcher, String quiltVersionNumber) {
    this.launcher = launcher;
    QuiltVersionNumber = quiltVersionNumber;
  }

  public File getQuiltVersionJson() {
    return new File(FileManger.getWdtcCache(), launcher.getVersionNumber() + "-quilt-" + QuiltVersionNumber + ".json");
  }

  public String getQuiltVersionJsonUrl() {
    return String.format(LibraryListUrl, launcher.getVersionNumber(), QuiltVersionNumber);
  }

  public JsonObject getQuiltGameVersionJsonObject() throws IOException {
    return JsonUtils.getJsonObject(getQuiltVersionJson());
  }

  @Override
  public String getModVersion() {
    return QuiltVersionNumber;
  }

  @Override
  public InstallTaskInterface getModInstallTask() {
    return new QuiltInstallTask(launcher, QuiltVersionNumber);
  }

  @Override
  public ModUtils.KindOfMod getModKind() {
    return ModUtils.KindOfMod.QUILT;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/quilt/QuiltDownloadInfo.kt
package org.wdt.wdtc.core.download.quilt

import com.google.gson.annotations.JsonAdapter
import org.wdt.utils.gson.readFileToJsonObject
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.*
import org.wdt.wdtc.core.manger.downloadSource
import org.wdt.wdtc.core.manger.wdtcCache
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.gson.DownloadInfoTypeAdapter
import java.io.File
import java.io.IOException

@JsonAdapter(DownloadInfoTypeAdapter::class)
open class QuiltDownloadInfo(
  protected val version: Version,
  override val modVersion: String
) : ModDownloadInfoInterface {
  private val libraryListUrl = "${downloadSource.quiltMetaUrl}/versions/loader/%s/%s/profile/json"

  constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) :
      this(version, versionsJsonObjectInterface.versionNumber!!)

  val quiltVersionJson: File
    get() = File(wdtcCache, "${version.versionNumber}-quilt-$modVersion.json")
  val quiltVersionJsonUrl: String
    get() = libraryListUrl.format(version.versionNumber, modVersion)

  @get:Throws(IOException::class)
  val quiltGameVersionJsonObject
    get() = quiltVersionJson.readFileToJsonObject()
  override val modInstallTask: ModInstallTaskInterface
    get() = QuiltInstallTask(version, modVersion)
  override val modKind: KindOfMod
    get() = KindOfMod.QUILT

  override fun toString(): String {
    return "QuiltDownloadInfo(modVersion='$modVersion')"
  }

}
